keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,365 @@
+| access | Read-Only | Read-Only|
+
+Podresource API looks like a good fit for this endpoint. Our major concern is how will this integration impact the shape of this API in the future.
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads. If kubelet is in the init phase and not all pod’s information is known, then the API should report that information. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but kubelet is working properly this API should return most recent data about local pods that are computed by kubelet even if those data were not reported or accepted by Kube-API. |
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
+The new API should return pods statuses that are computed by kubelet even though they are not reported back to kube-API but reflect the current pod’s state.
+
+#### Kubelet restarts issue
+
+Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) is impacting this API. After restart kubelet marks all pods as NotReady and then starts the init loop checking the actual pod's status. When kubelet is in this `unknown` state it should not serve the API data which might be incorrect. This API needs to be reliable and returned data should reflect the actual state of the pods. We don't want this API to return partial data. We should resolve this issue before the beta release. One of the possible fixes is returning error code indicating that kubelet is in the init phase and that pod's data are not known yet.
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `podinfo` in Kubelet, listening on a unix socket at `/var/lib/kubelet/pod-info/kubelet.sock`. The endpoint will be versioned.
```

Comment:
Maybe instead of referring to this as `podInfo` we should consider it `podStatus` or similar? Not sure on the name here, but `podInfo` is probably broader than we'd want here.

Reply:
podInfo is more flexible and we don't limit this API shape in this early phase.